### PR TITLE
Add Neural sparse tool, do refactor using AbstractRetrievalTool

### LINKED ISSUE #1675: [FEATURE] Tool development for NeuralSparse Tool
**Is your feature request related to a problem?**
Agent framework (https://github.com/opensearch-project/ml-commons/issues/1161 ) needs tools to extend functionality. This issue propose to add NeuralSparse tool to it.

**What solution would you like?**
Build a tool to support neural_sparse query. The interface should be similar to VectorDB tool. Since this tool is very similar to VectorDB tool, we can create an abstract retriever tool class and make the 2 tools extend it.
 
**What alternatives have you considered?**
N/A.

**Do you have any additional context?**
#1161 #1546

### PR DESCRIPTION
**Pull request by zhichao-aws**
### Description
Add NeuralSparseTool implementation. Implement an abstract class AbstractRetrievalTool, making NeuralSparseTool and VectorDBTool extend this class.
 
### Issues Resolved
#1675 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by zhichao-aws** [comment]
The failed IT on windows 11 seems already exists https://github.com/opensearch-project/ml-commons/actions/runs/6954022968/job/18920109040. Not introduced by this PR.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/AbstractRetrieverTool.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import lombok.Setter;
+import lombok.extern.log4j.Log4j2;
+
+/**
+ * Abstract tool supports search paradigms in neural-search plugin.
+ */
+@Log4j2
+@Getter
+@Setter
+public abstract class AbstractRetrieverTool implements Tool {
```
Comment (arjunkumargiri) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/AbstractRetrieverTool.java:39]: Thanks for doing the refactoring.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/AbstractRetrieverTool.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+                if (hits != null && hits.length > 0) {
+                    StringBuilder contextBuilder = new StringBuilder();
+                    for (int i = 0; i < hits.length; i++) {
+                        SearchHit hit = hits[i];
+                        String doc = AccessController.doPrivileged((PrivilegedExceptionAction<String>) () -> {
+                            Map<String, Object> docContent = new HashMap<>();
+                            docContent.put("_id", hit.getId());
+                            docContent.put("_source", hit.getSourceAsMap());
+                            return gson.toJson(docContent);
```
Comment (arjunkumargiri) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/AbstractRetrieverTool.java:102]: Let's also include _index and _score information to output.

**Review by arjunkumargiri — COMMENTED** [review]
Please create a example agent using Neural sparse tool and attach sample request/response details to PR.

**Review by dbwiddis — APPROVED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/AbstractRetrieverTool.java (hunk 2 above)
Comment (zhichao-aws) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/AbstractRetrieverTool.java]: Sure, makes sense

**Comment by zhichao-aws** [comment]
> Please create a example agent using Neural sparse tool and attach sample request/response details to PR.

Sure, will post one soon

**Comment by zhichao-aws** [comment]
- prerequisites: finsh the steps in the doc, get an sparse encoding index and neural sparse model id; a llm model id

## demo flow agent
register agent:
```
# request
POST /_plugins/_ml/agents/_register
{
  "name": "Test_Neural_Sparse_Agent_For_RAG",
  "type": "flow",
  "description": "this is a test agent",
  "tools": [
    {
      "type": "NeuralSparseTool",
      "parameters": {
        "model_id": "65fy8IsBzZuijRBEwRBX",
        "index": "sparse-index",
        "embedding_field": "sparse_embedding",
        "source_field": ["passage_text"],
        "input": "${parameters.question}"
      }
    },
    {
      "type": "MLModelTool",
      "description": "A general tool to answer any question",
      "parameters": {
        "model_id": "Gpct8YsBzZuijRBEkBIC",
        "prompt": "\n\nHuman:You are a professional data analysist. You will always answer question based on the given context first. If the answer is not directly shown in the context, you will analyze the data and find the answer. If you don't know the answer, just say don't know. \n\n Context:\n${parameters.NeuralSparseTool.output}\n\nHuman:${parameters.question}\n\nAssistant:"
      }
    }
  ]
}

# response
{
  "agent_id": "GQ4WH4wBIIa8ZwL88QuK"
}
```

execute agent:
```
# request
POST /_plugins/_ml/agents/GQ4WH4wBIIa8ZwL88QuK/_execute
{
    "parameters": {
        "question": "How many employee does company ABCD_TEST have?"
    }
}
# response
{
  "inference_results": [
    {
      "output": [
        {
          "result": " Based on the given context, company ABC_TEST has over 1000 employees."
        }
      ]
    }
  ]
}
```

## demo ReACT agent
register
```
# request
POST /_plugins/_ml/agents/_register
{
  "name": "Test_Agent_For_ReAct",
  "type": "conversational",
  "description": "this is a test agent",
  "llm": {
    "model_id": "Gpct8YsBzZuijRBEkBIC",
    "parameters": {
      "max_iteration": 5,
      "stop_when_no_tool_found": true,
      "response_filter": "$.completion"
    }
  },
  "memory": {
    "type": "conversation_index"
  },
  "tools": [
    {
      "type": "NeuralSparseTool",
      "name": "NeuralSparseSearchTool",
      "description": "A tool to search the knowledge base, you should always try to search data with this tool when encountering questions you don't know. Action Input: <natrual language question>",
      "parameters": {
        "model_id": "65fy8IsBzZuijRBEwRBX",
        "index": "hybrid-index",
        "embedding_field": "sparse_embedding",
        "source_field": [ "passage_text" ],
        "input": "${parameters.question}"
      }
    },
    {
      "type": "CatIndexTool",
      "name": "RetrieveIndexMetaTool",
      "description": "Use this tool to get OpenSearch index information: (health, status, index, uuid, primary count, replica count, docs.count, docs.deleted, store.size, primary.store.size)."
    }
  ],
  "app_type": "my app"
}
# response
{
  "agent_id": "Mw4bH4wBIIa8ZwL8ewvQ"
}
```
execute
```
# request
POST /_plugins/_ml/agents/Mw4bH4wBIIa8ZwL8ewvQ/_execute
{
  "parameters": {
    "question": "How many employee does company ABCD_TEST have?",
    "verbose": true
  }
}
# response
{
  "inference_results": [
    {
      "output": [
        {
          "name": "memory_id",
          "result": "Nw4bH4wBIIa8ZwL8zwtw"
        },
        {
          "name": "response",
          "result": "Thought: I do not have enough information to directly answer how many employees company ABCD_TEST has. I should use the NeuralSparseSearchTool to search for this information."
        },
        {
          "name": "response",
          "result": "\nObservation: \"{\\\"_index\\\":\\\"hybrid-index\\\",\\\"_source\\\":{\\\"passage_text\\\":\\\"The history of company ABC_TEST is over 100 years. The company has over 1000 employees.\\\"},\\\"_id\\\":\\\"IA4YH4wBIIa8ZwL8ywsd\\\",\\\"_score\\\":130.33943}\\n{\\\"_index\\\":\\\"hybrid-index\\\",\\\"_source\\\":{\\\"passage_text\\\":\\\"Company xyz have a history of 100 years.\\\"},\\\"_id\\\":\\\"3\\\",\\\"_score\\\":19.313019}\\n\""
        },
        {
          "name": "response",
          "result": "After searching through the knowledge base, I have found the answer."
        },
        {
          "name": "response",
          "result": "Company ABCD_TEST has over 1000 employees."
        }
      ]
    }
  ]
}
```

**Comment by zhichao-aws** [comment]
Hi @arjunkumargiri , I've resolved the comments. Please help review again

**Review by arjunkumargiri — APPROVED** [review]

**Comment by zhichao-aws** [comment]
@zane-neo @dbwiddis @arjunkumargiri Could you please help merge this PR? I don't have write access to this repo.

**Pull request merged**